packages/contracts-bedrock/scripts/checks/interfaces/main.go

```diff
@@ -0,0 +1,354 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sort"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+var excludeContracts = []string{
+	// External dependencies
+	"IERC20", "IERC721", "IERC721Enumerable", "IERC721Upgradeable", "IERC721Metadata",
+	"IERC165", "IERC165Upgradeable", "ERC721TokenReceiver", "ERC1155TokenReceiver",
+	"ERC777TokensRecipient", "Guard", "IProxy", "Vm", "VmSafe", "IMulticall3",
+	"IERC721TokenReceiver", "IProxyCreationCallback", "IBeacon",
+
+	// EAS
+	"IEAS", "ISchemaResolver", "ISchemaRegistry",
+
+	// TODO: Interfaces that need to be fixed
+	"IInitializable", "IPreimageOracle", "ILegacyMintableERC20", "IOptimismMintableERC20",
+	"IOptimismMintableERC721", "IOptimismSuperchainERC20", "MintableAndBurnable",
+	"KontrolCheatsBase", "IWETH", "IDelayedWETH", "IL2ToL2CrossDomainMessenger",
+	"ICrossL2Inbox", "ISystemConfigInterop", "IResolvedDelegateProxy",
+}
+
+type ContractDefinition struct {
+	ContractKind string `json:"contractKind"`
+	Name         string `json:"name"`
+}
+
+type ASTNode struct {
+	NodeType string   `json:"nodeType"`
+	Literals []string `json:"literals,omitempty"`
+	ContractDefinition
+}
+
+type ArtifactAST struct {
+	Nodes []ASTNode `json:"nodes"`
+}
+
+type Artifact struct {
+	AST ArtifactAST     `json:"ast"`
+	ABI json.RawMessage `json:"abi"`
+}
+
+func main() {
+	if err := run(); err != nil {
+		writeStderr("an error occurred: %v", err)
+		os.Exit(1)
+	}
+}
+
+func writeStderr(msg string, args ...any) {
+	_, _ = fmt.Fprintf(os.Stderr, msg+"\n", args...)
+}
+
+func run() error {
+	cwd, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("failed to get current working directory: %w", err)
+	}
+
+	artifactsDir := filepath.Join(cwd, "forge-artifacts")
+
+	artifactFiles, err := glob(artifactsDir, ".json")
+	if err != nil {
+		return fmt.Errorf("failed to get artifact files: %w", err)
+	}
+
+	// Remove duplicates from artifactFiles
+	uniqueArtifacts := make(map[string]string)
+	for contractName, artifactPath := range artifactFiles {
+		baseName := strings.Split(contractName, ".")[0]
+		uniqueArtifacts[baseName] = artifactPath
+	}
+
+	var hasErr int32
+	var outMtx sync.Mutex
+	fail := func(msg string, args ...any) {
+		outMtx.Lock()
+		writeStderr("❌  "+msg, args...)
+		outMtx.Unlock()
+		atomic.StoreInt32(&hasErr, 1)
+	}
+
+	sem := make(chan struct{}, runtime.NumCPU())
```

Comment:
I recommend the following package for things like this in the future: https://pkg.go.dev/golang.org/x/sync@v0.8.0/errgroup

If you look at the example code, it makes what you want to do really easy, you don't need to manage a semaphore or atomics

Reply:
Gotcha, I just copied another script but happy to switch to this